### PR TITLE
feat: add root.io support for container image scanning

### PR DESCRIPTION
**Pull request by knqyf263**
## Summary
- Add version constraint library supporting 5 operators (>, <, >=, <=, ==, \!=) for flexible version comparison
- Implement DriverProvider pattern for dynamic driver selection based on package information
- Add Root.io environment detection:
  - Debian/Ubuntu: packages with "root.io" in version
  - Alpine: packages with "roo7" in version
- Add constraint-based vulnerability detection using version ranges instead of single fixed versions
- Structure code to show VulnSrc interface should be implemented in trivy-db side
- Include comprehensive tests for all components

### Version Constraint Library (`pkg/detector/ospkg/version/`)
- `constraint.go`: Core constraint parsing and checking with 5 operators
- `compare.go`: Comparer interface with DEB and APK implementations
- Supports both comma and space separated constraints with OR logic
- Proper error handling for empty constraints/versions

### DriverProvider Pattern (`pkg/detector/ospkg/detect.go`)
- Added providers array that dynamically generate drivers based on package information
- Providers are tried before standard OS-specific drivers
- Clean separation between generic detection logic and specific implementations

### Root.io Scanner (`pkg/detector/ospkg/rootio/`)
- `provider.go`: Environment detection based on package version suffixes
- `rootio.go`: Main scanner with constraint-based vulnerability checking
- `vulnsrc.go`: Interface definition with mock implementation showing trivy-db integration pattern

### Common Driver Interface (`pkg/detector/ospkg/driver/`)
- Resolves circular import issues by providing common interface definitions
- Enables clean separation between packages

## Test Coverage
- Comprehensive test suite covering all components
- Tests for environment detection, constraint parsing, and vulnerability matching
- Mock implementations following actual trivy-db patterns

## Test plan
- [x] Unit tests pass for all new components
- [x] Integration with existing ospkg detection system
- [x] Proper error handling and edge cases covered
- [ ] Integration tests with actual Root.io container images (future work)
- [ ] Performance testing with large package sets (future work)

### CODE REVIEW

File: pkg/detector/ospkg/rootio/rootio.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			return false
+		} else if satisfied {
+			return true
+		}
+	}
+	return false
+}
+
+// IsSupportedVersion checks if the version is supported.
+// TODO: Should check if the base OS reaches EOL?
```
Comment (knqyf263) [inline pkg/detector/ospkg/rootio/rootio.go]: @chait-slim What do you think?

File: pkg/detector/ospkg/rootio/rootio.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+	return &Scanner{
+		comparer: comparer,
+		vs:       vs,
+		logger:   log.WithPrefix("rootio"),
+	}
+}
+
+// Detect vulnerabilities in package using Root.io scanner
+func (s *Scanner) Detect(ctx context.Context, osVer string, _ *ftypes.Repository, pkgs []ftypes.Package) ([]types.DetectedVulnerability, error) {
```
Comment (DmitriyLewen) [inline pkg/detector/ospkg/rootio/rootio.go:62]: I think make sense to add tests for this function

File: pkg/detector/ospkg/rootio/rootio.go (hunk 1 above)
Comment (benjifin) [reply pkg/detector/ospkg/rootio/rootio.go]: I don't think we have a strong need to check official EOL from our point of view - a use case of our users is using EOL software for longer thanks to the patches we create.

Comment (DmitriyLewen) [reply pkg/detector/ospkg/rootio/rootio.go]: So you create fixes for EOL OS versions (e.g. Alpine 3.18), right?

Comment (chait-slim) [reply pkg/detector/ospkg/rootio/rootio.go]: We support EOL distros, meaning we also create patches for them

**Comment by DmitriyLewen** [comment]
For test i changed version of `vim` in `alpine:3.18.7`.
https://github.com/DmitriyLewen/trivy-db/pkgs/container/trivy-db is used for test.
```
rootio:vim (alpine 3.18.7)
==========================
Total: 36 (UNKNOWN: 2, LOW: 5, MEDIUM: 24, HIGH: 5, CRITICAL: 0)

┌────────────┬────────────────┬──────────┬────────┬───────────────────┬─────────────────┬──────────────────────────────────────────────────────────────┐
│  Library   │ Vulnerability  │ Severity │ Status │ Installed Version │  Fixed Version  │                            Title                             │
├────────────┼────────────────┼──────────┼────────┼───────────────────┼─────────────────┼──────────────────────────────────────────────────────────────┤
│ libcrypto3 │ CVE-2024-6119  │ HIGH     │ fixed  │ 3.1.5-r0          │ 3.1.7-r0        │ openssl: Possible denial of service in X.509 name checks     │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2024-6119                    │
│            ├────────────────┼──────────┤        │                   ├─────────────────┼──────────────────────────────────────────────────────────────┤
│            │ CVE-2024-13176 │ MEDIUM   │        │                   │ 3.1.8-r0        │ openssl: Timing side-channel in ECDSA signature computation  │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2024-13176                   │
...
│            │                │          │        │                   │                 │                                                              │
├────────────┼────────────────┼──────────┤        ├───────────────────┼─────────────────┼──────────────────────────────────────────────────────────────┤
│ vim        │ CVE-2024-22667 │ HIGH     │        │ 9.0.2073-r00070   │ 9.0.2073-r00071 │ vim: Stack buffer over flow in did_set_langmap function in   │
│            │                │          │        │                   │                 │ map.c                                                        │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2024-22667                   │
│            ├────────────────┼──────────┤        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2023-48232 │ MEDIUM   │        │                   │                 │ vim: floating point exception in adjust_plines_for_skipcol() │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2023-48232                   │
│            ├────────────────┤          │        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2023-48233 │          │        │                   │                 │ vim: overflow with count for :s command                      │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2023-48233                   │
│            ├────────────────┤          │        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2023-48234 │          │        │                   │                 │ vim: overflow in nv_z_get_count                              │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2023-48234                   │
│            ├────────────────┤          │        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2023-48235 │          │        │                   │                 │ vim: overflow in ex address parsing                          │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2023-48235                   │
│            ├────────────────┤          │        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2023-48236 │          │        │                   │                 │ vim: overflow in get_number                                  │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2023-48236                   │
│            ├────────────────┤          │        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2023-48237 │          │        │                   │                 │ vim: buffer overflow in shift_line                           │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2023-48237                   │
│            ├────────────────┼──────────┤        │                   │                 ├──────────────────────────────────────────────────────────────┤
│            │ CVE-2024-43374 │ LOW      │        │                   │                 │ vim: use-after-free in alist_add() in src/arglist.c          │
│            │                │          │        │                   │                 │ https://avd.aquasec.com/nvd/cve-2024-43374                   │
...
└────────────┴────────────────┴──────────┴────────┴───────────────────┴─────────────────┴──────────────────────────────────────────────────────────────┘
``` 
For `vim` Trivy took vulns from `root.io`.
For `libcrypto3` Trivy took vulns from alpine DB.

**Review by DmitriyLewen — APPROVED** [review]

File: pkg/detector/ospkg/rootio/rootio_private_test.go
Comment (knqyf263) [inline pkg/detector/ospkg/rootio/rootio_private_test.go:1]: nit: I personally think it's fine to test private functions if really necessary, but as for `isVulnerable`, it seems like all cases can be covered in `constraint_test.go (TestConstraints_Check)` and `rootio_test.go (TestScanner_Detect)`. What do you think?

Comment (DmitriyLewen) [reply pkg/detector/ospkg/rootio/rootio_private_test.go:1]: I also thought about this.
But we can't handle case with combination of constraints (e.g. `"<1.0.0-2.root.io", ">=1.0.0-2 <1.0.0-3",`) `TestConstraints_Check`.
I wanted to add these test cases in `TestScanner_Detect`, but with each new test case it became more and more difficult to work with this test.
so I decided to make this separate file so that it would be easy to see/check whether the version is included in the constraints array
